lib/internal/event_target.js

```diff
@@ -0,0 +1,392 @@
+  const { once = false } = options;
+  if (typeof once !== 'boolean')
+    throw new ERR_INVALID_ARG_TYPE('options.once', 'boolean', once);
+  return { once };
+}
+
+function addEventListener(type, listener, options = {}) {
+  const state = targets.get(this);
+  if (state === undefined)
+    throw new ERR_INVALID_THIS('EventTarget');
+
+  validateType(type);
+  validateListener(listener);
+
+  const { once } = validateEventListenerOptions(options);
+
+  let handlers = state.events.get(type);
+  if (handlers === undefined) {
+    handlers = new Map();
+    state.events.set(type, handlers);
+  }
+
+  if (handlers.has(listener))
+    return;
+
+  const callback =
+    typeof listener === 'function' ?
+      listener.bind(this) :
+      listener.handleEvent.bind(listener);
+
+  const handler = {
+    callback,
+    once,
+    remove: removeEventListener.bind(this, type, listener)
+  };
+
+  handlers.set(listener, handler);
+}
+
+function removeEventListener(type, listener) {
+  const state = targets.get(this);
+  if (state === undefined)
+    throw new ERR_INVALID_THIS('EventTarget');
+
+  validateType(type);
+  validateListener(listener);
+
+  const handlers = state.events.get(type);
+  if (handlers !== undefined) {
+    handlers.delete(listener);
+    if (handlers.size === 0)
+      state.events.delete(type);
+  }
+}
+
+function addCatch(that, promise, event) {
+  const then = promise.then;
+  if (typeof then === 'function') {
+    then.call(promise, undefined, function(err) {
+      // The callback is called with nextTick to avoid a follow-up
+      // rejection from this promise.
+      process.nextTick(emitUnhandledRejectionOrErr, that, err, event);
+    });
+  }
+}
+
+function emitUnhandledRejectionOrErr(that, err, event) {
+  if (typeof that[kRejection] === 'function') {
+    that[kRejection](err, event);
+  } else {
+    process.emit('uncaughtException', err, event);
+  }
+}
+
+function dispatchEvent(event) {
+  const state = targets.get(this);
+  if (state === undefined)
+    throw new ERR_INVALID_THIS('EventTarget');
+
+  if (event == null ||
+      typeof event !== 'object' ||
+      typeof event.type !== 'string') {
+    throw new ERR_INVALID_ARG_TYPE('event', 'Event', event);
+  }
+
+  if (state.emitting.has(event.type)) {
+    throw new ERR_EVENT_RECURSION(event.type);
+  }
+
+  const handlers = state.events.get(event.type);
+  if (handlers !== undefined) {
+    state.emitting.add(event.type);
+    try {
+      const copy = ArrayFrom(handlers.values());
+      for (let n = 0; n < copy.length; n++) {
+        if (copy[n].once)
+          copy[n].remove();
+        try {
+          const result = copy[n].callback(event);
+          if (result !== undefined && result !== null) {
+            addCatch(this, result, event);
+          }
+        } catch (err) {
+          emitUnhandledRejectionOrErr(this, err, event);
+        }
+      }
+    } finally {
+      state.emitting.delete(event.type);
+    }
+  }
+}
+
+// EventEmitter-ish API:
```

Comment:
Note that just exposing these capabilities significantly changes the API surface of EventTarget. For example getting a listener count, event names and who's listening to what event invalidates various guarantees the DOM gives.

There is a whole (capability theory based) model of why exposing these is problematic IIRC. I am not saying we shouldn't expose these, but I am also saying that if we expose these extensions we should make sure with the WHATWG people that it's fine.

Reply:
In this case, given that we're being very explicit that this `EventTarget` diverges from the DOM `EventTarget`, and given the considerations around being at least somewhat compatible with `EventEmitter`, I'd rather we exposed these. Doing so is compatible with userland implementations (the signal in the `abort-controller` userland module, for instance, implements both EventEmitter and EventTarget APIs).

Reply:
Fair enough, maybe we should name it something that implies it extends an event target (eventhough it doesn't hold the same invariants so no dice for the LSP).

For reference here is the WHATWG discussion https://github.com/whatwg/dom/issues/412#issue-208246100

Reply:
I'll ask in the whatwg discussion to make sure it's fine, my intuition aligns with yours and I'd also prefer it if we exposed this if possible.

Reply:
It's a very bad idea to expose this; you're not supposed to be able to remove a listener that you didn't add.

Reply:
Note, I feel this way not because of the DOM whatsoever, but because exposing this basically means it's *never* safe to give out or expose an event emitter instance to anyone, ever, because they will always be able to break your assumptions about it

Reply:
@ljharb ... for DOM `EventTarget` and everything that entails, I would agree. the Node.js implementation, not exposing these additional functions makes the `EventTarget` far more likely to become a memory leak. I'd prefer to keep the hybrid EventTarget/EventEmitter behavior here with explicit recognition in the docs that the behavior (and requirements) are different here.

Reply:
The security/robustness considerations matter especially in node, where i might add a "cleanup memory leaks" event handler, and then some random code that gets access to the emitter removes it. Memory leaks are a *worse* problem with this exposed.

Reply:
Given that these already exist in `EventEmitter` and the idea here is to emulate `EventEmitter`, this is not making any of that any worse (or better) than it already is.

Reply:
Can you help me understand what's the benefit of shipping something called EventTarget that doesn't precisely match the browser's EventTarget?

Reply:
@jasnell fwiw I'm very torn on this and I see Jordan's point since this exposes the capability to remove an `abort` event listener from an added event (as an example). I think _in practice_ this is _good enough probably_ for users and this will allow for good interop but I also see the other point.

@ljharb I think the point would be to allow easy interop between EventEmitter and EventTarget - one example could be migrating a component from EventEmitter to EventTarget for browser interop where said component already uses EventEmitter APIs. This would allow gradual migration for instance.

Reply:
Just to show concrete code for the case I'm discussing:
```js
const controller = new AbortController();
const signal = controller.signal;
// some other place
const db = getDatabase();
signal.addEventListener('abort', () => {
  db.release();
});
// another consumer
signal.removeAllListeners('abort');
```

I don't think this is a particularly big footgun and it's not something users have been complaining about in general in Node.

It is also not in violation of the spec (as it extends the interface, nothing in the spec says an EventTarget can't implement a removeEventListeners) but it's certainly a violation of behavioral subtyping IMO.

I think it's _probably_ an acceptable violation but I think this is something we want to carefully consider.

(In any case Event and EventTarget aren't exposed here and this is marked experimental - should we open a new issue about this or have the discussion here? https://github.com/nodejs/node/issues/33528 )

Reply:
I will do the work to spin the discussion to another issue. Cheers.

Reply:
Opened an issue: https://github.com/nodejs/node/issues/33558

Reply:
> Note that just exposing these capabilities significantly changes the API surface of EventTarget. For example getting a listener count, event names and who's listening to what event invalidates various guarantees the DOM gives.
 
Sorry for the intervention but are these _guarantees_ codified anywhere, or, in other words, could theoretically relying on such be ever considered basically some sort of _bad habit_ instead?

> Can you help me understand what's the benefit of shipping something called EventTarget that doesn't precisely match the browser's EventTarget?

It may be worthwhile to consider if sticking to a definitely incomplete and asymmetrical, but  possibly also crippled and/or irrational implementation for _whatever_ reason is of any actual benefit.
There was some significant (but somewhat inconclusive) discussion on this over at WHATWG as well.
If anything, I'd expect the DOM approach to change (actually IIRC some engines already implement stuff like this under the hood, even if not exposing it).

Reply:
If the concern is interop, then only the *intersection* of features must be provided, not the union.

Reply:
It is a violation of the spec; the spec gives the precise members that a class can have.

I'd suggest if this is exposed then it's exposed somewhere besides the class (e.g. module-level).

Reply:
@domenic I see, so for example these methods could be exposed as static methods and not on `EventTarget` instances and that would be spec compliant?

@jasnell would that work? I think if Domenic and Jordan feel exposing these methods is counter-productive towards spec compliance (namely Domenic said it would be a violation) we should probably remove these methods from EventTarget and:

 - Either expose a `NodeEventTarget` interface that's a _subclass_ of `EventTarget` that adds these.
 - Expose static helper methods that perform these actions. Optionally those static methods would work for both EventTarget and EventEmitter.

What do you think?

Reply:
> Either expose a NodeEventTarget interface that's a subclass of EventTarget that adds these.

I'd prefer this approach.

Reply:
Sounds good to me.